runtime/src/accounts_cache.rs

```diff
@@ -0,0 +1,236 @@
+use dashmap::DashMap;
+use solana_sdk::{account::Account, clock::Slot, hash::Hash, pubkey::Pubkey};
+use std::{
+    collections::HashSet,
+    ops::Deref,
+    sync::{
+        atomic::{AtomicBool, AtomicU64, Ordering},
+        Arc, RwLock,
+    },
+};
+
+pub type SlotCache = Arc<SlotCacheInner>;
+
+#[derive(Default, Debug)]
+pub struct SlotCacheInner {
+    cache: DashMap<Pubkey, CachedAccount>,
+    same_account_writes: AtomicU64,
+    same_account_writes_size: AtomicU64,
+    unique_account_writes_size: AtomicU64,
+    is_frozen: AtomicBool,
+}
+
+impl SlotCacheInner {
+    pub fn report_slot_store_metrics(&self) {
+        datapoint_info!(
+            "slot_repeated_writes",
+            (
+                "same_account_writes",
+                self.same_account_writes.load(Ordering::Relaxed),
+                i64
+            ),
+            (
+                "same_account_writes_size",
+                self.same_account_writes_size.load(Ordering::Relaxed),
+                i64
+            ),
+            (
+                "unique_account_writes_size",
+                self.unique_account_writes_size.load(Ordering::Relaxed),
+                i64
+            )
+        );
+    }
+
+    pub fn insert(&self, pubkey: &Pubkey, account: Account, hash: Hash) {
+        if self.cache.contains_key(pubkey) {
+            self.same_account_writes.fetch_add(1, Ordering::Relaxed);
+            self.same_account_writes_size
+                .fetch_add(account.data.len() as u64, Ordering::Relaxed);
+        } else {
+            self.unique_account_writes_size
```

Comment:
I think unbounded cache size could be a dos attack vector. Consider 1000 tx on 10MB accounts per slot. :)
So, there should be some upperbound and forced flush state transition to mitigate that?

Well, AccountsDB even can't handle these writes, I guess. So, might not worth it.